pennylane_lightning/lightning_qubit.py

```diff
@@ -138,7 +116,10 @@ def apply(self, operations, rotations=None, **kwargs):
             self._pre_rotated_state = self._state
 
         if rotations:
-            self._state = self.apply_lightning(self._pre_rotated_state, rotations)
+            if any(isinstance(r, QubitUnitary)for r in rotations):
+                super().apply(operations=[], rotations=rotations)
+            else:
+                self._state = self.apply_lightning(np.copy(self._pre_rotated_state), rotations)
```

Comment:
> Being able to pass arbitrary unitaries to the backend would also allow us to directly support QubitUnitary

For gates that need to be defined by a unitary that can't be known ahead of time, passing in the unitary directly is fine.

> also new gates like QFT.

AFAIK, the QFT matrix is just parameterised by its dimension, so I'm not sure I see the need to pass in the matrix directly.

In general, if we can get away with an optimised kernel for gate application, we should. But that's only easily possible for gates that we know ahead of time.

Reply:
> For gates that need to be defined by a unitary that can't be known ahead of time, passing in the unitary directly is fine.

Nice! To clarify, this isn't currently supported right? At least, I get an error if I try:
```
TypeError: apply(): incompatible function arguments. The following argument types are supported:
E           1. (arg0: numpy.ndarray[complex128], arg1: List[str], arg2: List[List[int]], arg3: List[List[float]], arg4: int) -> None
```
Since we require `params` to be composed of lists of floats.

> AFAIK, the QFT matrix is just parameterised by its dimension, so I'm not sure I see the need to pass in the matrix directly.
>
> In general, if we can get away with an optimised kernel for gate application, we should. But that's only easily possible for gates that we know ahead of time.

Agreed! I guess what I'm getting at is that we just added `QFT` to PennyLane, and there will inevitably be a gap between gates supported in PL and gates with a defined matrix in the backend. It would be nice for us to have a pipeline to just support _any_ gate that has a matrix method, and we can optionally add in the kernel to the backend if we think it's a popular gate.

Reply:
> To clarify, this isn't currently supported right

Yes, that is correct. In theory, it could be made to work with the current pybind bindings, by encoding the matrix in the `vector<double> params` as a flattened array of doubles, but there's little benefit of doing so; modifying the bindings is a trivial thing to do.

> It would be nice for us to have a pipeline to just support any gate that has a matrix method, and we can optionally add in the kernel to the backend if we think it's a popular gate.

Agreed. Conceivably, we could just delegate any unrecognised gate type to the supplied matrix (and error if that's not provided). However, I am wary about always needing to pay the cost of generating the matrix and passing it to the backend, if we don't have to. I don't expect it to be a large penalty (unless we end up supporting controlled unitary operations with arbitrarily many controls), but I'd rather we do without it if possible.

So I think that at the end of the day, what we really want is not to pass the numpy array of the matrix to the backend directly, but rather to pass a functional that when invoked, generates the numpy array for the matrix. This will allow us to lazily generate the matrix as required.